Sources/PrimerSDK/Classes/Core/Primer/DependencyInjection.swift

```diff
@@ -36,36 +40,47 @@ final internal class DependencyContainer {
 
     private func register<T>(_ dependency: T) {
         let key = String(describing: T.self)
-        dependencies[key] = dependency as AnyObject
+        Self.queue.async(flags: .barrier) {
+            self.dependencies[key] = dependency as AnyObject
+        }
     }
 
     private func resolve<T>() -> T {
         let key = String(describing: T.self)
-        let dependency = dependencies[key] as? T
-
-        if dependency == nil {
-            if key == String(describing: AppStateProtocol.self) {
-                let appState: AppStateProtocol = AppState()
-                DependencyContainer.register(appState)
-                return self.resolve()
-
-            } else if key == String(describing: PrimerSettingsProtocol.self) {
-                let primerSettings: PrimerSettingsProtocol = PrimerSettings()
-                DependencyContainer.register(primerSettings)
-                return self.resolve()
-
-            } else if key == String(describing: PrimerThemeProtocol.self) {
-                let primerTheme: PrimerThemeProtocol = PrimerTheme()
-                DependencyContainer.register(primerTheme)
-                return self.resolve()
-            }
+
+        if let dependency = Self.queue.sync(execute: { dependencies[key] as? T }) {
+            return dependency
         }
 
-        precondition(
-            dependency != nil,
-            "No dependency found for \(key)! must register a dependency before resolve."
-        )
+        return Self.queue.sync {
+            if let dependency = self.dependencies[key] as? T {
+                return dependency
+            }
 
-        return dependency!
+            let dependency: T? = self.createDependency(for: key)
+            if let dependency = dependency {
+                self.dependencies[key] = dependency as AnyObject
+            }
+
+            precondition(
+                dependency != nil,
+                "No dependency found for \(key)! must register a dependency before resolve."
+            )
+
+            return dependency!
```

Comment:
force unwrap alert

Reply:
Yep, I dont like it either. But it existed before (red line 69), and the precondition should protect us.
To make this return a `T?` would be quite ugly at the callsites I suppose 😔

Reply:
I see it is safe because of the precondition and it fails anyway, and we want it to fail here, I just don't like force unwrapping. `preconditionFailure` looks like a better solution. WDYT?